crates/pathfinder/src/state/contract_hash.rs

```diff
@@ -0,0 +1,414 @@
+    #[serde(deny_unknown_fields)]
+    pub struct Program<'a> {
+        #[serde(skip_serializing_if = "Vec::is_empty")]
+        pub attributes: Vec<serde_json::Value>,
+
+        pub builtins: Vec<Cow<'a, str>>,
+
+        pub data: Vec<Cow<'a, str>>,
+
+        #[serde(borrow)]
+        pub debug_info: Option<&'a serde_json::value::RawValue>,
+
+        // Important that this is ordered by the numeric keys, not lexicographically
+        pub hints: BTreeMap<u64, Vec<serde_json::Value>>,
+
+        pub identifiers: serde_json::Value,
+
+        #[serde(borrow)]
+        pub main_scope: Cow<'a, str>,
+
+        // Unlike most other integers, this one is hex string. We don't need to interpret it,
+        // it just needs to be part of the hashed output.
+        #[serde(borrow)]
+        pub prime: Cow<'a, str>,
+
+        pub reference_manager: serde_json::Value,
+    }
+
+    #[cfg(test)]
+    mod roundtrip_tests {
+        // FIXME: we should have many test cases utilizing this.
+        #[allow(unused)]
+        fn roundtrips<'a, T>(input: &'a str)
+        where
+            T: serde::Deserialize<'a> + serde::Serialize,
+        {
+            use super::super::PythonDefaultFormatter;
+
+            let parsed: T = serde_json::from_str(input).unwrap();
+            let mut ser =
+                serde_json::Serializer::with_formatter(Vec::new(), PythonDefaultFormatter);
+            parsed.serialize(&mut ser).unwrap();
+            let bytes = ser.into_inner();
+            let output = std::str::from_utf8(&bytes).expect("serde does this unchecked");
+
+            // these need to be byte for byte equal because we hash this
+            assert_eq!(input, output);
+        }
+    }
+
+    #[cfg(test)]
+    mod test_vectors {
+        #[tokio::test]
+        async fn first() {
+            // this test is a bit on the slow side because of the download and because of the long
+            // processing time in dev builds. expected --release speed is 100 contracts/s.
+            let expected = pedersen::StarkHash::from_hex_str(
+                "0031da92cf5f54bcb81b447e219e2b791b23f3052d12b6c9abd04ff2e5626576",
+            )
+            .unwrap();
+
+            // this is quite big payload, ~500kB
+            let resp = reqwest::get("https://external.integration.starknet.io/feeder_gateway/get_full_contract?contractAddress=0x4ae0618c330c59559a59a27d143dd1c07cd74cf4e5e5a7cd85d53c6bf0e89dc")
+                .await
+                .unwrap();
+
+            let payload = resp.text().await.expect("response wasn't a string");
+
+            // for bad urls the response looks like:
+            // 500
+            // {"code": "StarknetErrorCode.UNINITIALIZED_CONTRACT", "message": "Contract with address 2116724861677265616176388745625154424116334641142188761834194304782006389228 is not deployed."}
+
+            let hash = super::super::compute_contract_hash(&payload).unwrap();
+
+            assert_eq!(hash, expected);
```

Comment:
This is the costly first test, keep, #[ignore] or delete?

Reply:
How costly is it in time? We have some longer (10s) tests already with Infura, so it might just be fine?

Reply:
Yes, you don't notice it when running all tests. But it's bad form to lean on ext services for testing.